yt_dlp/extractor/acfun.py

```diff
@@ -0,0 +1,192 @@
+import re
+
+from .common import InfoExtractor
+from ..utils import (
+    float_or_none,
+    int_or_none,
+    traverse_obj,
+)
+
+
+class AcFunVideoIE(InfoExtractor):
+    _VALID_URL = r'(?x)https?://(?:www\.acfun\.cn/v/)ac(?P<id>[_\d]+)'
+
+    _TESTS = [{
+        'url': 'https://www.acfun.cn/v/ac35457073',
+        'info_dict': {
+            'id': '35283657',
+            'title': '【十五周年庆】文章区UP主祝AcFun生日快乐！',
+            'duration': 455.21,
+            'timestamp': 1655289827,
+        },
+        'params': {
+            # m3u8 download
+            'skip_download': True,
+        },
+        'skip': 'Geo-restricted to China',
+    }]
+
+    def parse_format(self, info, video_id):
+        vcodec, acodec = None, None
+
+        codec_str = info.get('codecs') or ''
+        m = re.match('(?P<vc>[^,]+),(?P<ac>[^,]+)', codec_str)
+        if m:
+            vcodec = m.group("vc")
+            acodec = m.group("ac")
+
+        formats, _ = self._extract_m3u8_formats_and_subtitles(
+            info['url'], video_id)
+
+        # it seems AcFun do not have subtitles
+
+        return {
+            **formats[0],
+            'fps': int_or_none(info.get('frameRate')),
+            'width': int_or_none(info.get('width')),
+            'height': int_or_none(info.get('height')),
+            'vcodec': vcodec,
+            'acodec': acodec,
+            'tbr': float_or_none(info.get('avgBitrate')),
+            'ext': 'mp4'
+        }
+
+    def parse_format_list(self, jobj, video_id):
+        formats = []
+        for video in jobj:
+            fmt = self.parse_format(video, video_id)
+            formats.append(fmt)
+
+        self._sort_formats(formats)
+
+        return formats
+
+    def gen_other_video_info_map(self, video_info):
+        return {
+            'duration': float_or_none(video_info.get('durationMillis'), 1000),
+            'timestamp': int_or_none(video_info.get('uploadTime'), 1000),
+        }
```

Comment:
It loops to me like all these three should be merged into a single function

Reply:
ok